### ISSUE TITLE
materialized_views_test.TestMaterializedViews.test_populate_mv_after_insert_wide_rows is flaky

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#15046]
Saw in https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3045/testReport/junit/materialized_views_test/TestMaterializedViews/Tests___Sanity_Tests___test_populate_mv_after_insert_wide_rows/

I couldn't reproduce this locally.

@fruch is this one of the new tests, added to gating?

**Comment by mykaul** [comment]
@fruch ?

**Comment by fruch** [comment]
history of this test seem like it was failing with this for a while, and not it's failing with a different error:
```
>           raise AssertionError('Critical errors found: {}\nOther errors: {}'.format(
                critical_errors, found_errors))
E           AssertionError: Critical errors found: [('node1', ['scylla: seastar/include/seastar/core/gate.hh:144: future<> seastar::gate::close(): Assertion `!_stopped && "seastar::gate::close() cannot be called more than once"\' failed.', 'Aborting on shard 0.']), ('node2', ['scylla: seastar/include/seastar/core/gate.hh:144: future<> seastar::gate::close(): Assertion `!_stopped && "seastar::gate::close() cannot be called more than once"\' failed.', 'Aborting on shard 0.']), ('node3', ['scylla: seastar/include/seastar/core/gate.hh:144: future<> seastar::gate::close(): Assertion `!_stopped && "seastar::gate::close() cannot be called more than once"\' failed.', 'Aborting on shard 0.'])]
E           Other errors: []
```

@bhalevy are you aware of those type of failures ?

meanwhile let's take it from gating

**Comment by bhalevy** [comment]
> history of this test seem like it was failing with this for a while, and not it's failing with a different error:
> 
> ```
> >           raise AssertionError('Critical errors found: {}\nOther errors: {}'.format(
>                 critical_errors, found_errors))
> E           AssertionError: Critical errors found: [('node1', ['scylla: seastar/include/seastar/core/gate.hh:144: future<> seastar::gate::close(): Assertion `!_stopped && "seastar::gate::close() cannot be called more than once"\' failed.', 'Aborting on shard 0.']), ('node2', ['scylla: seastar/include/seastar/core/gate.hh:144: future<> seastar::gate::close(): Assertion `!_stopped && "seastar::gate::close() cannot be called more than once"\' failed.', 'Aborting on shard 0.']), ('node3', ['scylla: seastar/include/seastar/core/gate.hh:144: future<> seastar::gate::close(): Assertion `!_stopped && "seastar::gate::close() cannot be called more than once"\' failed.', 'Aborting on shard 0.'])]
> E           Other errors: []
> ```
> 
> @bhalevy are you aware of those type of failures ?

Apparently these are due to issue in https://github.com/scylladb/scylladb/pull/15213 branch

**Comment by fruch** [comment]
this happened 5 times, since 10 Aug:
https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/98ae1610-4b1b-11ee-b96a-256b7e8b2261

I think it's enough time to take it out of the gate

**Comment by mykaul** [comment]
@fruch - is it a recent addition to gating?

**Comment by fruch** [comment]
> @fruch - is it a recent addition to gating?

it was added to the gating in the big batch added in a month ago: https://github.com/scylladb/scylla-dtest/pull/3452

**Comment by nyh** [comment]
I looked a the test. My guess is that this is a (very minor) **Scylla bug**, not a test problem:

This test (test_populate_mv_after_insert_wide_rows) does:
1. It creates a cluster with 3 nodes, RF=1.
2. It then builds a table with 5 partitions of 1000 rows each. 
3. It then adds a view to this table, waits for it to be built.
4. Then it for each of the 5000 rows, it reads it from the view, asserting it's there.

The good news is that all of this passes successfully... The view gets built and read just fine.
The only problem is that the test finds a message in the log:
```
ERROR 2023-08-14 07:32:54,996 [shard 0] view - Error applying view update to 127.0.53.3 (view: ks.t_by_v, base token: -4069959284402364209, view token: 8209703677368372043): exceptions::mutation_write_failure_exception (Operation failed for ks.t_by_v - received 0 responses and 1 failures from 1 CL=ONE.
```

This error must have been a transient error during the view building, as eventually we were able to read all the data so the view building must have finished.

The error mentions tokens, which can be translated to keys id=1 and v=34:

```
>>> from cassandra.metadata import Murmur3Token
>>> Murmur3Token.hash_fn(int.to_bytes(1, length=4))
-4069959284402364209

>>> for i in range(10000):
...     if Murmur3Token.hash_fn(int.to_bytes(i, length=4)) == 8209703677368372043:
...         print(i)
... 
34
>>> Murmur3Token.hash_fn(int.to_bytes(34, length=4))
8209703677368372043
```

I don't know what is special in this particular row and can't find the Scylla log file (@fruch, can you help me) to see if there are any other error messages that happened during the view build. In particular, this error message is rate limited - so it might be the first of 100 messages (e.g. all ids between 0 and 100, an 34 just happened to be printed first). Or not, I don't know.

I suspect that the view building on node 1 experienced a transient problem which prevented the view update from being sent to the view replica on node 3 (note that because N=3, RF=1, view replicas are typically on a different node on this test). I can't be sure why it failed without finding the logs of node 3 (@fruch maybe you can help me), but I have a guess, that will need to proved or discredited: 

View building starts on node 1 as soon as node 1 learns that the CREATE MATERIALIZED VIEW has happened. However, it is conceivable that when node 1 starts to send view updates, and decides it should send some of them to node 3, node 3 doesn't yet know about this view's existence so the write fails to node 3. The view building will later try again, and by that time node 3 will have received the updated schema, and the second write will succeed and the view building will proceed.

If this explanation is correct, we can modify the view building so that it doesn't start until all live nodes are aware of the new view. Perhaps waiting for schema agreement is the right way to go.
Alternatively, we should consider not reporting errors during view build as "ERROR"s, since the view building retries anyway, and these should be considered INFO at most, if at all.

CC @cvybhu @eliransin

**Comment by fruch** [comment]
we don't have the log of all those occurrences any more, we don't persist all of the jobs for too long, and in dtest case we don't move them anywhere else.

@benipeled can we enlarge the dtest gating job history a bit ?

meanwhile there are similar tests with MVs that having similar failures here's one:
https://jenkins.scylladb.com/job/scylla-master/job/dtest-release-with-raft/245/testReport/junit/concurrent_schema_changes_test/TestConcurrentSchemaChanges/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split010___test_create_lots_of_mv_concurrently/

and a link to the dashboard show similar ones (i.e. that have the same error mentioned in thier failure):
https://70f106c98484448dbc4705050eb3f7e9.us-east-1.aws.found.io:9243/goto/557f8e00-4d96-11ee-b96a-256b7e8b2261

**Comment by nyh** [comment]
> meanwhile there are similar tests with MVs that having similar failures here's one: https://jenkins.scylladb.com/job/scylla-master/job/dtest-release-with-raft/245/testReport/junit/concurrent_schema_changes_test/TestConcurrentSchemaChanges/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split010___test_create_lots_of_mv_concurrently/

Thanks, this one indeed looks similar. This one also has an error in node1 trying to send view updates to node2.
The node1 log https://jenkins.scylladb.com/job/scylla-master/job/dtest-release-with-raft/245/artifact/logs-full.release.010/1693892699459_concurrent_schema_changes_test.py%3A%3ATestConcurrentSchemaChanges%3A%3Atest_create_lots_of_mv_concurrently/node1.log has:
```
INFO  2023-09-05 05:43:45,833 [shard 0:stre] schema_tables - Creating lots_o_views.src_by_c2 id=2e01bd50-4baf-11ee-988f-c6c02f6a2b6d version=dae806bf-19ea-3f75-99c8-037d25edb1d6
INFO  2023-09-05 05:43:45,835 [shard 0:stre] view - Building view lots_o_views.src_by_c2, starting at token minimum token
ERROR 2023-09-05 05:43:45,836 [shard 0:stre] view - Error applying view update to 127.0.75.2 (view: lots_o_views.src_by_c2, base token: -8978909661410770097, view token: 1287679541648494871): exceptions::mutation_write_failure_exception (Operation failed for lots_o_views.src_by_c2 - received 0 responses and 1 failures from 1 CL=ONE.)
WARN  2023-09-05 05:43:45,836 [shard 0:stre] view - Error executing build step for base lots_o_views.source_data: exceptions::mutation_write_failure_exception (Operation failed for lots_o_views.src_by_c2 - received 0 responses and 1 failures from 1 CL=ONE.)
INFO  2023-09-05 05:43:47,016 [shard 0:stre] view - Finished building view lots_o_views.src_by_c2
```

As I assumed, the view building on node started 2 milliseconds after the MV was created on node 1, `2023-09-05 05:43:45,835`. Just a millisecond later, the first write to node 2 failed. Looking at the node 2 log https://jenkins.scylladb.com/job/scylla-master/job/dtest-release-with-raft/245/artifact/logs-full.release.010/1693892699459_concurrent_schema_changes_test.py%3A%3ATestConcurrentSchemaChanges%3A%3Atest_create_lots_of_mv_concurrently/node2.log we see:
```
INFO  2023-09-05 05:43:45,829 [shard 0:stat] migration_manager - Create new view: org.apache.cassandra.config.CFMetaData@0x600001130400[cfId=2e01bd50-4baf-11ee-988f-c6c02f6a2b6d,
INFO  2023-09-05 05:43:45,834 [shard 0:stre] schema_tables - Altering lots_o_views.source_data id=1fc41c10-4baf-11ee-908b-f711dd34e11d version=aaedc4a9-12ec-3263-b5d9-c91e3f2e11f9
INFO  2023-09-05 05:43:45,834 [shard 0:stre] schema_tables - Creating lots_o_views.src_by_c2 id=2e01bd50-4baf-11ee-988f-c6c02f6a2b6d version=dae806bf-19ea-3f75-99c8-037d25edb1d6
WARN  2023-09-05 05:43:45,836 [shard 0:stat] storage_proxy - Failed to apply mutation from 127.0.75.1#0: data_dictionary::no_such_column_family (Can't find a column family with UUID 2e01bd50-4baf-11ee-988f-c6c02f6a2b6d)
```

I think this confirms my guess - node 2 receives view updates two milliseconds after learning of the view, and apparently it's not yet fully aware of this very recently-added table and fails to find it.

**Comment by nyh** [comment]
As this is a (minor) Scylla bug, not a dtest flakiness, I'll change the issue title. I'll include here the failed dtest names for searchability:
* `materialized_views_test.py`::`TestMaterializedViews`::`test_populate_mv_after_insert_wide_rows`
* `concurrent_schema_changes_test.py`::`TestConcurrentSchemaChanges`::`test_create_lots_of_mv_concurrently`

**Comment by mykaul** [comment]
@nyh - the retry mechanism is based on hints, btw? I think it's very interesting to figure out the reason for the transient failure - maybe will give us a clue wrt inconsistent MVs.

**Comment by nyh** [comment]
I think I explained above the reason for the transient failure - node 1 started view building of pre-existing data, sent a view update to node 2 - just before node 2 fully learned about the view existing so it refuses the write.
IIRC there are no hints involved here - view building remembers its position and what it needs to resend. Also note that the view is fully consistent at the end - the only "bug" here is the log spam.

**Comment by bhalevy** [comment]
> I think I explained above the reason for the transient failure - node 1 started view building of pre-existing data, sent a view update to node 2 - just before node 2 fully learned about the view existing so it refuses the write. IIRC there are no hints involved here - view building remembers its position and what it needs to resend. Also note that the view is fully consistent at the end - the only "bug" here is the log spam.

Maybe in this particular case, but I'm concerned about not properly synchronizing with schema changes.
See https://github.com/scylladb/scylladb/issues/15048#issuecomment-1677241939
(and the issues look duplicated)

**Comment by fruch** [comment]
if so @nyh, can change those log print into a warning ? so we won't need to filter them all over the place ?

**Comment by mykaul** [comment]
@fruch - remove from gating please. Lets continue the dissection and analysis in parallel.

**Comment by fruch** [comment]
> @fruch - remove from gating please. Lets continue the dissection and analysis in parallel.

it's all the materialized view tests, that are gonna be taken out of gating, what mentioned here can happen on any of them

**Comment by mykaul** [comment]
> > @fruch - remove from gating please. Lets continue the dissection and analysis in parallel.
> 
> it's all the materialized view tests, that are gonna be taken out of gating, what mentioned here can happen on any of them

We haven't had those failures before - or have we?
We haven't changed the logging, to the best of my knowledge.
If it's a benign issue - even better, we can fix it on the side branch, then re-apply either the test, the fix or both and get them back into gating. Right now, it's (yet another) source for gating flakiness.

**Comment by benipeled** [comment]
> we don't have the log of all those occurrences any more, we don't persist all of the jobs for too long, and in dtest case we don't move them anywhere else.
> 
> @benipeled can we enlarge the dtest gating job history a bit ?
> 

Artifacts archived for 240 days - should be enough in this case but feel free to adjust it as needed if you think it's required
https://github.com/scylladb/scylla-dtest/blob/next/vars/dtestPipeline.groovy#L59-L60

**Comment by hshanghai** [comment]
您好，您的邮件我已经收到，我会尽快和你回复（自动回复）

**Comment by fruch** [comment]
> > we don't have the log of all those occurrences any more, we don't persist all of the jobs for too long, and in dtest case we don't move them anywhere else.
> > @benipeled can we enlarge the dtest gating job history a bit ?
> 
> Artifacts archived for 240 days - should be enough in this case but feel free to adjust it as needed if you think it's required https://github.com/scylladb/scylla-dtest/blob/next/vars/dtestPipeline.groovy#L59-L60

@benipeled and the jobs themself ? I think in the rate we have for it, they doesn't last long
https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4176/ is from 9 days ago.
we need much more jobs history on that one, I would say 3 months.

**Comment by fruch** [comment]
@benipeled 

https://github.com/scylladb/scylla-dtest/pull/3567 it was by number of jobs, and the day for the artifacts doesn't really helps, since we keep the links to the job, and not to specific artifacts.

**Comment by nyh** [comment]
> We haven't had those failures before - or have we? We haven't changed the logging, to the best of my knowledge. If it's a benign issue - even better, we can fix it on the side branch, then re-apply either the test, the fix or both and get them back into gating. Right now, it's (yet another) source for gating flakiness.

I think it's perhaps not a common occurrence, the view build on the first node races with the second node learning that the view exists. Perhaps the schema-changing usually wins the race. I don't know why it has become (if at all) more common than it was in the past. Moreover, this problem only applies to tests that do view building (adding a view to a table with pre-existing data) - many MV tests add data to the view after it's created and are not vulnerable to this issue.

**Comment by benipeled** [comment]
> > > we don't have the log of all those occurrences any more, we don't persist all of the jobs for too long, and in dtest case we don't move them anywhere else.
> > > @benipeled can we enlarge the dtest gating job history a bit ?
> > 
> > 
> > Artifacts archived for 240 days - should be enough in this case but feel free to adjust it as needed if you think it's required https://github.com/scylladb/scylla-dtest/blob/next/vars/dtestPipeline.groovy#L59-L60
> 
> @benipeled and the jobs themself ? I think in the rate we have for it, they doesn't last long 

We can keep the job themself for long time but it means console only which is not enough

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4176/ is from 9 days ago. we need much more jobs history on that one, I would say 3 months.

What about this build? it has all artifacts

**Comment by fruch** [comment]
> > > > we don't have the log of all those occurrences any more, we don't persist all of the jobs for too long, and in dtest case we don't move them anywhere else.
> > > > @benipeled can we enlarge the dtest gating job history a bit ?
> > > 
> > > 
> > > Artifacts archived for 240 days - should be enough in this case but feel free to adjust it as needed if you think it's required https://github.com/scylladb/scylla-dtest/blob/next/vars/dtestPipeline.groovy#L59-L60
> > 
> > 
> > @benipeled and the jobs themself ? I think in the rate we have for it, they doesn't last long
> 
> We can keep the job themself for long time but it means console only which is not enough
> 
> https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4176/ is from 9 days ago. we need much more jobs history on that one, I would say 3 months.
> 
> What about this build? it has all artifacts

yeah, when there no other errors causing it not to have it :)

**Comment by fruch** [comment]
> @fruch - did we take this out of gating? @nyh - is it reasonable to change this ERROR to WARN?

yes, cause of those  kind of issues, all of those test we taken out of gating
https://github.com/scylladb/scylla-dtest/commit/17f9d015868ffb6fb27595099ce63dcc502832c7

**Comment by fruch** [comment]
we are seeing this a lot in SCT as well:
```
2023-09-30 06:39:50.379 <2023-09-30 06:39:50.000>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=04f44e1e-ae85-4253-9ac6-f6f3556af219: type=DATABASE_ERROR regex=(^ERROR|!\s*?ERR).*\[shard.*\] line_number=6245 node=longevity-large-partitions-200k-pks-db-node-d62e8bac-0-3
2023-09-30T06:39:50+00:00 longevity-large-partitions-200k-pks-db-node-d62e8bac-0-3      !ERR | scylla[6843]:  [shard 11:stre] view - (rate limiting dropped 26 similar messages) Error applying view update to 10.142.0.111 (view: scylla_bench.view_test, base token: 3520969294134539810, view token: 8659755778285643669): exceptions::unavailable_exception (Cannot achieve consistency level for cl ONE. Requires 1, alive 0)
```

repeating @mykaul, can we change those print level ?, or we need to a add code to ignore all over the place ?

**Comment by fruch** [comment]
I hundreds of those errors prints in a case which mostly do topology changes on a multi-dc cluster:

```
2023-10-02 17:01:44.225 <2023-10-02 17:01:44.000>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=ec00c543-bb0f-484d-9800-bc383f0c5f15: type=DATABASE_ERROR regex=(^ERROR|!\s*?ERR).*\[shard.*\] line_number=2187 node=parallel-topology-schema-changes-mu-db-node-7162315c-3
2023-10-02T17:01:44+00:00 parallel-topology-schema-changes-mu-db-node-7162315c-3      !ERR | scylla[5623]:  [shard 0:stat] view - Error applying view update to 10.4.9.255 (view: keyspace1.standard1_c3_nemesis_index, base token: 8687565012191682492, view token: -783736015806138877): data_dictionary::no_such_column_family (Can't find a column family with UUID 5d06aba1-6145-11ee-9092-ffeb434a7910)
```

when should we expect those error ?, it's not just on index creation, and seems like it might happen any time a new is added. or a node is stopped.


## Installation details

Kernel Version: 5.15.0-1045-aws
Scylla version (or git commit hash): `5.4.0~dev-20231002.1640f83fdc31` with build-id `d751f4ab981bda6f045bb902f0d341e61f0ac3a7`





Cluster size: 12 nodes (i3en.2xlarge)

Scylla Nodes used in this run:
- parallel-topology-schema-changes-mu-db-node-7162315c-9 (18.169.241.143 | 10.3.8.151) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-8 (18.134.14.223 | 10.3.11.54) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-7 (35.178.231.250 | 10.3.8.101) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-6 (3.249.50.46 | 10.4.11.196) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-5 (63.33.195.147 | 10.4.10.90) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-4 (34.252.175.149 | 10.4.8.128) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-3 (34.253.133.72 | 10.4.9.255) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-2 (54.194.157.249 | 10.4.8.112) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-17 (54.246.26.71 | 10.4.11.5) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-16 (63.35.215.119 | 10.4.10.195) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-15 (18.130.145.101 | 10.3.10.253) (shards: -1)
- parallel-topology-schema-changes-mu-db-node-7162315c-14 (54.216.171.70 | 10.4.9.118) (shards: -1)
- parallel-topology-schema-changes-mu-db-node-7162315c-13 (52.30.33.55 | 10.4.11.229) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-12 (3.10.144.217 | 10.3.9.121) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-11 (18.170.55.64 | 10.3.11.217) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-10 (18.130.5.161 | 10.3.9.174) (shards: 7)
- parallel-topology-schema-changes-mu-db-node-7162315c-1 (3.250.46.244 | 10.4.9.254) (shards: 7)


OS / Image: `ami-0000f82a29037b49d ami-0205476ccb2b99643` (aws: undefined_region)

Test: `longevity-multidc-schema-topology-changes-12h-test`
Test id: `7162315c-e5dc-41a6-940d-13e7997f7162`
Test name: `scylla-master/longevity/longevity-multidc-schema-topology-changes-12h-test`
Test config file(s):

- [longevity-multidc-parallel-topology-schema-changes-12h.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/b295d74a58ef552a7af6a6520468a548c24453a9/test-cases/longevity/longevity-multidc-parallel-topology-schema-changes-12h.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor 7162315c-e5dc-41a6-940d-13e7997f7162`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=7162315c-e5dc-41a6-940d-13e7997f7162)
- Show all stored logs command: `$ hydra investigate show-logs 7162315c-e5dc-41a6-940d-13e7997f7162`


## Logs:
- **db-cluster-7162315c.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/db-cluster-7162315c.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/db-cluster-7162315c.tar.gz)
- **sct-runner-events-7162315c.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/sct-runner-events-7162315c.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/sct-runner-events-7162315c.tar.gz)
- **sct-7162315c.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/sct-7162315c.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/sct-7162315c.log.tar.gz)
- **loader-set-7162315c.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/loader-set-7162315c.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/loader-set-7162315c.tar.gz)
- **monitor-set-7162315c.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/monitor-set-7162315c.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/7162315c-e5dc-41a6-940d-13e7997f7162/20231003_171218/monitor-set-7162315c.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-multidc-schema-topology-changes-12h-test/15/)
[Argus](https://argus.scylladb.com/test/4d9b162e-c35a-41e6-a0ac-d677a64c26ac/runs?additionalRuns[]=7162315c-e5dc-41a6-940d-13e7997f7162)
</details>

**Comment by kostja** [comment]
@nyh view updates should be signed with schema version just like normal writes. And schema version fencing should work. Why doesn't it?

**Comment by dani-tweig** [comment]
@nyh 
Do you think it would be enough to change the log level to WARN? or do we need to investigate what @kostja recommended, to find out why version fencing didnt work this time?

**Comment by nyh** [comment]
I don't know what Kostja meant about schema versions and fencing (he may be right, but I'm just not experienced enough in this area), so I'll use simpler words: 

It is pointless for any node to start its **view building** process before it is sure that all nodes know of the new view, because this node will surely need to send updates to every other nodes - and it will fail if the other nodes don't know on the view building.

So I think the real fix is to delay the start of view building a bit - until schema agreement, and I think we should keep this open until we do this.

I don't know what is the right "process" to wait until all the nodes know of the new view, and whether we should wait for every single node (in any case, view building is likely to pause if it can't reach one of the nodes, so the view building process does depend on no node being down).

**Comment by nyh** [comment]
> > @fruch - remove from gating please. Lets continue the dissection and analysis in parallel.
> 
> it's all the materialized view tests, that are gonna be taken out of gating, what mentioned here can happen on any of them

I missed that this was done. In my analysis above I thought that only a couple of specific tests had this problem - not all of them... So,... Now it turns out for a few months (since September), **all** of the materialized views tests have been off gating. That's not good. So I guess we need to fix this issue once and for all to get the tests back to gating!
I see that @kostja assigned this issue to @wmitros?

In comment https://github.com/scylladb/scylladb/issues/15046#issuecomment-1932642793 I suggested what I think the "best" thing to do would be - to delay view building, but perhaps a simpler and good-enough fix would be to downgrade the ERROR to WARN, considering that the view building is already coded to ignore these errors and try again (which is why it works correctly, besides printing those scary errors).

**Comment by piodul** [comment]
@mykaul they were not put back to gating yet.

https://github.com/scylladb/scylladb/pull/15640 downgraded the error message to WARN so it no longer should cause failures. However, as right now we are trying to stabilize CI, it might be a good idea to wait with putting it back to gating until CI becomes more stable.

**Comment by piodul** [comment]
> [#15640](https://github.com/scylladb/scylladb/pull/15640) downgraded the error message to WARN so it no longer should cause failures.

I removed the `symptom/ci stability` label, while this issue still exists it should not cause issues in CI.

**Comment by mykaul** [comment]
@piodul - in theory, it should go back to gating then.
In practice, it is still unstable - see https://jenkins.scylladb.com/job/scylla-master/job/dtest-debug/464/artifact/logs-full.debug.004/ (dtest-gw1) : 
```
22:12:03,727 630     dtest_setup                    DEBUG    dtest_setup.py      :61   | test_populate_mv_after_insert_wide_rows: Allocated cluster ID 2: /jenkins/workspace/scylla-master/dtest-debug/scylla/.dtest/dtest-kp3vdib3
22:12:03,794 630     ccm                            DEBUG    cluster.py          :757  | test_populate_mv_after_insert_wide_rows: node1: data_center=dc1 rack=RAC1 snitch=org.apache.cassandra.locator.GossipingPropertyFileSnitch
22:12:03,874 630     ccm                            DEBUG    cluster.py          :757  | test_populate_mv_after_insert_wide_rows: node2: data_center=dc1 rack=RAC1 snitch=org.apache.cassandra.locator.GossipingPropertyFileSnitch
22:12:03,955 630     ccm                            DEBUG    cluster.py          :757  | test_populate_mv_after_insert_wide_rows: node3: data_center=dc1 rack=RAC1 snitch=org.apache.cassandra.locator.GossipingPropertyFileSnitch
22:12:03,997 630     ccm                            DEBUG    cluster.py          :757  | test_populate_mv_after_insert_wide_rows: start_nodes: no_wait=False wait_for_binary_proto=True wait_other_notice=True wait_normal_token_owner=True force_wait_for_cluster_start=True
22:12:04,020 630     ccm                            DEBUG    cluster.py          :757  | test_populate_mv_after_insert_wide_rows: node1: Starting scylla: args=['/jenkins/workspace/scylla-master/dtest-debug/scylla/.dtest/dtest-kp3vdib3/test/node1/bin/scylla', '--options-file', '/jenkins/workspace/scylla-master/dtest-debug/scylla/.dtest/dtest-kp3vdib3/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--abort-on-internal-error', '1', '--api-address', '127.0.2.1', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--blocked-reactor-notify-ms', '5000', '--overprovisioned', '--prometheus-address', '127.0.2.1', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
22:12:07,204 630     ccm                            INFO     cluster.py          :760  | test_populate_mv_after_insert_wide_rows: node1: Started scylla: pid: 899
22:12:07,231 630     ccm                            DEBUG    cluster.py          :757  | test_populate_mv_after_insert_wide_rows: node2: Starting scylla: args=['/jenkins/workspace/scylla-master/dtest-debug/scylla/.dtest/dtest-kp3vdib3/test/node2/bin/scylla', '--options-file', '/jenkins/workspace/scylla-master/dtest-debug/scylla/.dtest/dtest-kp3vdib3/test/node2/conf/scylla.yaml', '--log-to-stdout', '1', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--abort-on-internal-error', '1', '--api-address', '127.0.2.2', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--blocked-reactor-notify-ms', '5000', '--overprovisioned', '--prometheus-address', '127.0.2.2', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
22:12:13,718 630     ccm                            INFO     cluster.py          :760  | test_populate_mv_after_insert_wide_rows: node2: Started scylla: pid: 933
22:12:13,745 630     ccm                            DEBUG    cluster.py          :757  | test_populate_mv_after_insert_wide_rows: node3: Starting scylla: args=['/jenkins/workspace/scylla-master/dtest-debug/scylla/.dtest/dtest-kp3vdib3/test/node3/bin/scylla', '--options-file', '/jenkins/workspace/scylla-master/dtest-debug/scylla/.dtest/dtest-kp3vdib3/test/node3/conf/scylla.yaml', '--log-to-stdout', '1', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--abort-on-internal-error', '1', '--api-address', '127.0.2.3', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--blocked-reactor-notify-ms', '5000', '--overprovisioned', '--prometheus-address', '127.0.2.3', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
22:12:22,387 630     ccm                            INFO     cluster.py          :760  | test_populate_mv_after_insert_wide_rows: node3: Started scylla: pid: 951
22:12:23,061 630     cassandra.policies             INFO     policies.py         :307  | test_populate_mv_after_insert_wide_rows: Using datacenter 'dc1' for DCAwareRoundRobinPolicy (via host '127.0.2.1:9042'); if incorrect, please specify a local_dc to the constructor, or limit contact points to local cluster nodes
22:12:23,061 630     cassandra.cluster              INFO     dtest_setup.py      :497  | test_populate_mv_after_insert_wide_rows: New Cassandra host <Host: 127.0.2.3:9042 dc1> discovered
22:12:23,062 630     cassandra.cluster              INFO     dtest_setup.py      :497  | test_populate_mv_after_insert_wide_rows: New Cassandra host <Host: 127.0.2.2:9042 dc1> discovered
22:12:23,103 630     dtest_class                    DEBUG    dtest_class.py      :211  | test_populate_mv_after_insert_wide_rows: CREATE KEYSPACE ks WITH replication={'class':'NetworkTopologyStrategy', 'replication_factor':1}
22:12:23,215 630     cassandra.protocol             WARNING  libevreactor.py     :371  | test_populate_mv_after_insert_wide_rows: Server warning: Tables in this keyspace will be replicated using Tablets and will not support CDC, LWT and counters features. To use CDC, LWT or counters, drop this keyspace and re-create it without tablets by adding AND TABLETS = {'enabled': false} to the CREATE KEYSPACE statement.
22:12:23,215 630     cassandra.protocol             WARNING  libevreactor.py     :371  | test_populate_mv_after_insert_wide_rows: Server warning: Using Replication Factor replication_factor=1 lower than the minimum_replication_factor_warn_threshold=3 is not recommended.
22:13:47,448 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:683  | test_populate_mv_after_insert_wide_rows: Waiting for view ks.t_by_v to finish building: num_nodes=3 num_alive=3 cl=QUORUM
22:13:47,459 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}})
22:13:47,459 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node 3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1/127.0.2.1 is not done yet
22:13:48,469 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'STARTED': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1')}, 'SUCCESS': {UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}})
22:13:48,469 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node 3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1/127.0.2.1 is not done yet
22:13:49,476 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'STARTED': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1')}, 'SUCCESS': {UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}})
22:13:49,477 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node 3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1/127.0.2.1 is not done yet
22:13:50,482 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'STARTED': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1')}, 'SUCCESS': {UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}})
22:13:50,483 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node 3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1/127.0.2.1 is not done yet
22:13:51,488 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:51,488 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:13:52,494 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:52,494 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:13:53,498 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:53,498 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:13:54,505 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:54,505 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:13:55,511 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:55,511 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:13:56,516 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:56,516 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:13:57,521 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:57,521 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:13:58,525 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:58,525 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:13:59,532 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:13:59,533 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:14:00,537 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:14:00,537 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:14:01,541 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}, 'STARTED': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd')}})
22:14:01,542 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:697  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: node f2becccb-d26a-4a2a-9237-cacbe32f78dd/127.0.2.2 is not done yet
22:14:02,546 630     tools.tables_view_manager      DEBUG    tables_view_manager.py:692  | test_populate_mv_after_insert_wide_rows: wait_for_view ks.t_by_v: status=defaultdict(<class 'set'>, {'SUCCESS': {UUID('f2becccb-d26a-4a2a-9237-cacbe32f78dd'), UUID('3b3cd54d-f094-4ba1-b2b2-e069f7cd1fb1'), UUID('5c14729c-f4ad-4927-8fda-22e65865cc3b')}})
22:15:02,153 630     errors                         ERROR    conftest.py         :223  | test_populate_mv_after_insert_wide_rows: test failed: 
self = <materialized_views_test.TestMaterializedViews object at 0x7f4b034b89b0>

    @unmark.next_gating  # https://github.com/scylladb/scylladb/issues/15046
    def test_populate_mv_after_insert_wide_rows(self):
        """Test that a view is OK when created with existing data with wide rows"""
    
        session = self.prepare()
    
        session.execute("CREATE TABLE t (id int, v int, PRIMARY KEY (id, v))")
    
        for i in range(5):
            for j in range(10000):
                session.execute(f"INSERT INTO t (id, v) VALUES ({i}, {j})")
    
        session.execute("CREATE MATERIALIZED VIEW t_by_v AS SELECT * FROM t WHERE v IS NOT NULL AND id IS NOT NULL PRIMARY KEY (v, id)")
    
        wait_for_view(cluster=self.cluster, session=session, ks="ks", view="t_by_v")
    
        for i in range(5):
            for j in range(10000):
>               assert_one(session, f"SELECT * FROM t_by_v WHERE id = {i} AND v = {j}", [j, i])

materialized_views_test.py:1576: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

session = <cassandra.cluster.Session object at 0x7f4b02c03620>
query = 'SELECT * FROM t_by_v WHERE id = 3 AND v = 1124', expected = [1124, 3]
cl = None

    def assert_one(session, query, expected, cl=None):
        """
        Assert query returns one row.
        @param session Session to use
        @param query Query to run
        @param expected Expected results from query
        @param cl Optional Consistency Level setting. Default ONE
    
        Examples:
        assert_one(session, "LIST USERS", ['cassandra', True])
        assert_one(session, query, [0, 0])
        """
        simple_query = SimpleStatement(query, consistency_level=cl)
        res = session.execute(simple_query)
        list_res = _rows_to_list(res)
>       assert list_res == [expected], f"Expected {[expected]} from {query}, but got {list_res}"
E       AssertionError: Expected [[1124, 3]] from SELECT * FROM t_by_v WHERE id = 3 AND v = 1124, but got []

tools/assertions.py:147: AssertionError
22:15:02,153 630     errors                         ERROR    conftest.py         :223  | test_populate_mv_after_insert_wide_rows: test failed: 
self = <materialized_views_test.TestMaterializedViews object at 0x7f4b034b89b0>

    @unmark.next_gating  # https://github.com/scylladb/scylladb/issues/15046
    def test_populate_mv_after_insert_wide_rows(self):
        """Test that a view is OK when created with existing data with wide rows"""
    
        session = self.prepare()
    
        session.execute("CREATE TABLE t (id int, v int, PRIMARY KEY (id, v))")
    
        for i in range(5):
            for j in range(10000):
                session.execute(f"INSERT INTO t (id, v) VALUES ({i}, {j})")
    
        session.execute("CREATE MATERIALIZED VIEW t_by_v AS SELECT * FROM t WHERE v IS NOT NULL AND id IS NOT NULL PRIMARY KEY (v, id)")
    
        wait_for_view(cluster=self.cluster, session=session, ks="ks", view="t_by_v")
    
        for i in range(5):
            for j in range(10000):
>               assert_one(session, f"SELECT * FROM t_by_v WHERE id = {i} AND v = {j}", [j, i])

materialized_views_test.py:1576: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

session = <cassandra.cluster.Session object at 0x7f4b02c03620>
query = 'SELECT * FROM t_by_v WHERE id = 3 AND v = 1124', expected = [1124, 3]
cl = None

    def assert_one(session, query, expected, cl=None):
        """
        Assert query returns one row.
        @param session Session to use
        @param query Query to run
        @param expected Expected results from query
        @param cl Optional Consistency Level setting. Default ONE
    
        Examples:
        assert_one(session, "LIST USERS", ['cassandra', True])
        assert_one(session, query, [0, 0])
        """
        simple_query = SimpleStatement(query, consistency_level=cl)
        res = session.execute(simple_query)
        list_res = _rows_to_list(res)
>       assert list_res == [expected], f"Expected {[expected]} from {query}, but got {list_res}"
E       AssertionError: Expected [[1124, 3]] from SELECT * FROM t_by_v WHERE id = 3 AND v = 1124, but got []
```

**Comment by mykaul** [comment]
@piodul - it seem to have happened again - see https://jenkins.scylladb.com/job/scylla-master/job/dtest-release/936/testReport/junit/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_populate_mv_after_insert_wide_rows/ - let's look at this and try to understand if the test is still unstable, if it's tablets, or whatnot.

**Comment by lsfreitas** [comment]
Seen: https://jenkins.scylladb.com/job/scylla-master/job/dtest-release/938/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_populate_mv_after_insert_wide_rows/

[node1.log](https://github.com/user-attachments/files/20850270/node1.log)
[node2.log](https://github.com/user-attachments/files/20850271/node2.log)
[node3.log](https://github.com/user-attachments/files/20850272/node3.log)

cc @piodul @wmitros

**Comment by Annamikhlin** [comment]
seen also on https://jenkins.scylladb.com/job/scylla-master/job/dtest-release/955/testReport/junit/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split000___test_populate_mv_after_insert/

**Comment by lsfreitas** [comment]
Seen in dtest debug: https://jenkins.scylladb.com/job/scylla-master/job/dtest-debug/513/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split019___test_populate_mv_after_insert_wide_rows/

[dtest-gw2.log](https://github.com/user-attachments/files/21505219/dtest-gw2.log)
[node1.log](https://github.com/user-attachments/files/21505222/node1.log)
[node2.log](https://github.com/user-attachments/files/21505221/node2.log)
[node3.log](https://github.com/user-attachments/files/21505223/node3.log)

**Comment by mykaul** [comment]
@mlitvk - is https://github.com/scylladb/scylladb/pull/25171 relevant for this one?

**Comment by mlitvk** [comment]
> [@mlitvk](https://github.com/mlitvk) - is [#25171](https://github.com/scylladb/scylladb/pull/25171) relevant for this one?

I think the issues are not related.

I see this issue started as being about view building running when some node is not aware of the view yet, resulting in ERROR logs. If I understand correctly, this was benign and didn't result in base-view inconsistencies, and was solved by reducing the log level.

But more recently the problem with the test is that it fails on another issue of base-view inconsistency.
I see we get these errors:
```
WARN  2025-07-27 09:14:18,524 [shard 0:strm] view - Could not find e67f2c6a-15ec-43d6-8ec9-36cb556d163b in base_endpoints=[eaca0526-036e-4e6e-99ca-022a3f9824cf]
```
and there is tablet migration:
```
INFO  2025-07-27 09:14:02,962 [shard 0: gms] raft_topology - Initiating tablet streaming (migration) of d3b39da0-6ac9-11f0-8231-0934a1fa3ec4:7 to eaca0526-036e-4e6e-99ca-022a3f9824cf:0

INFO  2025-07-27 09:14:06,823 [shard 0: gms] raft_topology - Initiating tablet cleanup of d3b39da0-6ac9-11f0-8231-0934a1fa3ec4:7 on e67f2c6a-15ec-43d6-8ec9-36cb556d163b:0
```

I think this is similar to https://github.com/scylladb/scylladb/issues/17603 which should be fixed by the view build coordinator.

**Comment by wmitros** [comment]
Right, the recent failures are not reproductions of this issue - I see we actually had the issue for them https://github.com/scylladb/scylladb/issues/23931 and we came to the same conclusion there but we started posting new failures here for some reason.
This issue about error-level log spam was already fixed by reducing the log-level. I see we also had a discussion about whether these logs should appear in the first place, but from what I'm seeing, in the recent runs (the posted failures) we don't get even warn-level  "Error applying view update" messages (we get messages related to view updates failed due to the tablet migration). The failed runs use tablets and I don't know where we're running tests without tablets, so I ran the test with vnodes myself and there's also no warn messages, so we might have implemented this suggestion as well in one of many fixes added since this issue was created.
So I will close this issue - we sent https://github.com/scylladb/scylladb/pull/15640 to fix it and there have been no actual reproductions of it. If we get more failures with missing rows, we can post them on https://github.com/scylladb/scylladb/issues/23931
